pkg/event/stats.go

```diff
@@ -24,3 +22,32 @@ type DataMetrics struct {
 	Labels        prometheus.Labels
 	Value         float64
 }
+
+// DataDetails .. details for data
+type DataDetails struct {
+	IFace        string
+	State        PTPState
+	ClockType    ClockType
+	Metrics      map[ValueType]DataMetric
+	time         int64
+	logData      string
+	signalSource EventSource // GNSS PPS
+	sourceLost   bool
+}
+
+func (d *Data) updateState() {
+	state := PTP_UNKNOWN
+	for _, detail := range d.Details { // 2 ts2phc or 2 dpll etc
+		switch detail.State {
+		case PTP_FREERUN: // if its free ru and main state is not holdover then this is the state
+			state = detail.State
```

Comment:
if main state is holdever  and this detail.State is the last one in the slice then the main state will be updated to freerun which is not correct.

Reply:
Added stats_test  unit test and fixed it